### PR TITLE
ComponentHandleState event

### PR DESCRIPTION
**Pull request by gradientvera**
This is a directed component event. What this means is that this will only be dispatched directed to the component and its references. (I changed the enumerator method so it'll be dispatched to the component AFTER the references)
This allows us to handle component states from entity systems, therefore removing *even more* logic from component, and moving us closer to struct components, maybe.

The reason I haven't coded an event for getting the state is because that change wouldn't be backwards-compatible, we'd have to change EVERY SINGLE COMPONENT'S GET COMPSTATE to use the event instead.

Anyway, **do NOT merge without @Acruid and @PJB3005's explicit approvals.**

### CODE REVIEW

**Comment by Acruid** [comment]
Is this a good idea? The component states are technically part of networking, they are not game logic. Component structs *can* have methods and properties on them to help access the data they contain.

**Comment by gradientvera** [comment]
Mmmm, how would component structs handling states look? Would they use an interface or something?

File: Robust.Shared/GameObjects/Component.cs [hunk 1]
```diff
@@ -228,6 +228,7 @@ public virtual ComponentState GetComponentState(ICommonSession player)
         }
 
         /// <inheritdoc />
+        [Obsolete("Use the ComponentHandleState event instead.")]
```
Comment (PJB3005) [inline Robust.Shared/GameObjects/Component.cs]: Probably a good idea to curb your enthusiasm on this and *not* mark something like this obsolete yet.

File: Robust.Shared/GameObjects/EntityEventBus.Directed.cs [hunk 2]
```diff
@@ -294,9 +294,17 @@ public void Dispose()
...
+            /// </summary>
             private IEnumerable<Type> GetReferences(Type type)
             {
-                return _comFac.GetRegistration(type).References;
+                var list = _comFac.GetRegistration(type).References;
+
+                for (var i = list.Count-1; i >= 0; i--)
+                {
+                    yield return list[i];
+                }
```
Comment (PJB3005) [inline Robust.Shared/GameObjects/EntityEventBus.Directed.cs]: `IComponentRegistration` makes no guarantee that the first element is the type itself. Filter to ignore the type in the inner loop and then do a final `yield return` at the end instead.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: Robust.Shared/GameObjects/Component.cs (hunk 1 above)
Comment (gradientvera) [reply Robust.Shared/GameObjects/Component.cs]: My YEAST is SLOW.
Done.

File: Robust.Shared/GameObjects/EntityEventBus.Directed.cs (hunk 2 above)
Comment (gradientvera) [reply Robust.Shared/GameObjects/EntityEventBus.Directed.cs]: Done, that's indeed better

**Review by PJB3005 — APPROVED** [review]

**Pull request merged**